### ISSUE TITLE
future reproduce plan for PV-RCNN refinement layer

### ISSUE THREAD

**Issue by muzi2045** [jhultman/PV-RCNN#10]
I have tried to reproduce the propasal layer training(same as SECOND), and got some reasonable result in my own dataset, it look not bad.(the truck yaw angle looks like a little offset)
![pv-rcnn](https://user-images.githubusercontent.com/12984898/76137902-0b00c600-607d-11ea-82e5-959ba605c79b.PNG)
now I am working on Nuscenes dataset api and reproduce the refinement layer result in PV-RCNN with multi-class dataset.

But here still exist some problem:
1. multi-class training loss are not stable decline(maybe it's the personal dataset problem).
2. the inference part need to be refactory because you have set special score threshold for every detection class.
3. change pickle to json or messagepack to operate on serialize, because json is the same as pickle but faster.
4. Add dataset Eval scripts.

have you tried train on refinement layer and got some results that are make sense?
@jhultman

**Comment by jhultman** [comment]
Hi, thanks so much for your comments. It makes me very happy to see other people working on this project.

1. > multi-class training loss are not stable decline(maybe it's the personal dataset problem).

I have not put much thought into optimizer parameter choices so maybe they are poorly chosen? Also in Det3D repository, they use FastAI [implementation](https://github.com/poodarchu/Det3D/blob/master/det3d/solver/learning_schedules_fastai.py#L77) of OneCycleLR scheduler, whereas I use master branch PyTorch version to avoid so much copy-paste code. I need to check to make sure there are no key implementation differences here. Also this repo currently does not ignore targets with too few lidar points which might result in a few very challenging examples.

2. > the inference part need to be refactory because you have set special score threshold for every detection class.

I will work on this. I think it can be implemented in a fast and simple vectorized way by indexing with `class_idx`. I am also working on batch inference for proposal stage since I think it will be important for end-to-end training of proposal and refinement stage. One challenge is how to handle the varying number of boxes per class and per batch.

3. > change pickle to json or messagepack to operate on serialize, because json is the same as pickle but faster.

I wasn't aware of this but I just did some research and you are completely right, json can easily be over 10x faster. I don't mind switching to json serialization. Do you mean for annotations or sample augmentation database or both? Does read/write take a long time when starting training for NuScenes dataset?

4. > Add dataset Eval scripts.

I am still trying to decide the best way to add eval support. I have tried some preliminary KITTI eval code adapted from [here](https://github.com/traveller59/kitti-object-eval-python) but I am hesitant to add it to this repo since the code is quite difficult to understand and makes a lot of assumptions about how the annotations are formatted. I imagine eval for NuScenes dataset will be easy and clean to implement since their devkit already handles the mAP calculations.

5. > have you tried train on refinement layer and got some results that are make sense? 

I have been focusing on proposal stage for the past couple of days, but since that part now seems to be relatively bug-free I am eager to begin working on refinement stage again. I still need to implement two parts that are missing:
- Predicted keypoint weighting (Section 3.2).
- IoU-guided scoring (Section 3.3).

I also may need to rewrite the RefinementTargetAssigner since I may have broken it with the latest few commits.

**Comment by muzi2045** [comment]
First, thanks for your reply!
> 1. > change pickle to json or messagepack to operate on serialize, because json is the same as pickle but faster.
> 
> I wasn't aware of this but I just did some research and you are completely right, json can easily be over 10x faster. I don't mind switching to json serialization. Do you mean for annotations or sample augmentation database or both? Does read/write take a long time when starting training for NuScenes dataset?

Yes, it will cost a lot of time in annotations generate when you are handle huge dataset such as Nuscenes. Both write and read are slow.

**Comment by jhultman** [comment]
> Yes, it will cost a lot of time in annotations generate when you are handle huge dataset such as Nuscenes. Both write and read are slow.

So I spent some time looking into this and compared read and write speeds for pickle, json, [hickle](https://github.com/telegraphic/hickle), and [h5py](https://github.com/h5py/h5py). I ran benchmarks on KITTI annotations and KITTI augmentation database using HDD (not SSD). I also made a pseudo-dataset 10x as large as KITTI to simulate what results would be like for larger datasets.

Here are some timings (in seconds) on the KITTI 10x dataset:

For train annotations (37,120 keyframes):
```
Pickle dump calibobj: 1.74
Pickle load calibobj: 0.89
Pickle dump calibobj high_protocol: 1.74
Pickle load calibobj high_protocol: 0.89
Pickle dump calibdict: 1.41
Pickle load calibdict: 0.61
json dump numpyencoder: 9.67
json load numpyencoder: 2.87
json dump no-overhead: 8.50
h5py dump: 0.95
```

For augmentation database (154,540 boxes and corresponding point clouds):
```
h5py dump: 2.69
h5py dump no-vlen: 0.93
pickle dump: 1.22
```

I am not an expert with these packages so I may have missed some possible optimizations, but as preliminary results it seems latest pickle (Python 3.7+) is actually quite competitive for this particular dataset and use-case. Also some of these alternatives are quite restrictive (json cannot pickle numpy data so requires converters, h5py cannot handle multidimensional variable-length arrays so requires reshaping before write and after read, h5py cannot write raw python dicts). I would strongly consider `hickle` since it is a single-line drop-in replacement for `pickle` but it seems to be extremely slow at least in my experiments.

**Comment by jhultman** [comment]
I can share more details if you're curious. I don't have NuScenes dataset on my development machine so I can't test that right now.

**Comment by muzi2045** [comment]
that's interesting...
My test environment is python3.6, maybe the pickle has been speed up in Python3.7
I'll check the whole data read process to find which place cost time most.

**Comment by jhultman** [comment]
Hi, regarding (2) I have just pushed the batch inference code with different score thresholds per class.

**Comment by jhultman** [comment]
Regarding 4), I plan to add eval code for KITTI shortly.

**Comment by muzi2045** [comment]
it looks like the official implement of PV-RCNN will be there 
[PointCloudDet3D](https://github.com/sshaoshuai/PointCloudDet3D)
but it used many original second.pytorch repo's code, and the depends are old.

**Comment by jhultman** [comment]
> it looks like the official implement of PV-RCNN will be there
> [PointCloudDet3D](https://github.com/sshaoshuai/PointCloudDet3D)
> but it used many original second.pytorch repo's code, and the depends are old.

Thanks, I saw that and I think it will be a useful reference. But the original second.pytorch code is quite complex and hard to debug/extend so I hope to provide a clean implementation here.

**Comment by muzi2045** [comment]
https://github.com/skyhehe123/SA-SSD
the network in this repo looks like PV-RCNN, but the submodule only for training not for inference.
so it will be faster than PV-RCNN(two-stage).

**Comment by jhultman** [comment]
Hi, thanks for link. I will look at that repo and consider implementing it here. Also, sorry for the lack of updates in a while. I have been debugging SECOND training to make sure there are no bugs in training code. I hope to push the changes soon.

**Comment by cqsnove** [comment]
> Hi, thanks for link. I will look at that repo and consider implementing it here. Also, sorry for the lack of updates in a while. I have been debugging SECOND training to make sure there are no bugs in training code. I hope to push the changes soon.

Thank you for your work and look forward to your updates
